src/client/datascience/jupyter/kernels/kernelSelections.ts

```diff
@@ -317,8 +329,40 @@ export class KernelSelectionProvider {
         @inject(IFileSystem) private readonly fs: IFileSystem,
         @inject(IPathUtils) private readonly pathUtils: IPathUtils,
         @inject(IKernelFinder) private readonly kernelFinder: IKernelFinder,
-        @inject(IPythonExtensionChecker) private readonly extensionChecker: IPythonExtensionChecker
-    ) {}
+        @inject(IPythonExtensionChecker) private readonly extensionChecker: IPythonExtensionChecker,
+        @inject(IDisposableRegistry) disposableRegistry: IDisposableRegistry,
+
+        @inject(IJupyterSessionManagerFactory) private jupyterSessionManagerFactory: IJupyterSessionManagerFactory
+    ) {
+        disposableRegistry.push(
+            this.jupyterSessionManagerFactory.onRestartSessionCreated(this.addKernelToIgnoreList.bind(this))
+        );
+        disposableRegistry.push(
+            this.jupyterSessionManagerFactory.onRestartSessionUsed(this.removeKernelFromIgnoreList.bind(this))
+        );
+    }
+
+    /**
+     * Ensure kernels such as those associated with the restart session are not displayed in the kernel picker.
+     *
+     * @param {Kernel.IKernelConnection} kernel
+     * @memberof KernelSelector
+     */
+    public addKernelToIgnoreList(kernel: Kernel.IKernelConnection): void {
```

Comment:
Moved from KernelSelector class to this so that it can be used by NativeNotebooks (this class is already used by native notebooks as it doesn't have UI specific code)